js/src/setup-mc/setup-stepper/setup-free-listings/index.js

```diff
@@ -62,8 +70,40 @@ const SetupFreeListings = ( props ) => {
 		return {};
 	};
 
-	const handleSubmitCallback = () => {
-		onContinue();
+	const handleSubmitCallback = async ( values ) => {
+		/**
+		 * Even though we already have auto-save effects in the FormContent,
+		 * we are saving the form values here again to be sure,
+		 * because the auto-save may not be fired
+		 * when users were having focus on the text input fields
+		 * and then immediately click on the Continue button.
+		 */
+		const {
+			shipping_country_rates: shippingRatesValue,
+			...settingsValue
+		} = values;
+
+		setSaving( true );
+
+		try {
+			await Promise.all( [
+				saveSettings( settingsValue ),
+				saveShippingRates( shippingRatesValue ),
+			] );
+			onContinue();
+		} catch ( error ) {
+			createNotice(
+				'error',
+				__(
+					'There is a problem in saving your settings and shipping rates. Please try again later.',
+					'google-listings-and-ads'
+				)
+			);
+		}
+
+		if ( isMounted() ) {
```

Comment:
❔  I'm curious about what this deals with?

Reply:
If the users go to other page (e.g. by clicking on the step number at the top) while the saving is in progress, there will be a console error like "setState() on an unmounted component", so we check for `isMounted()` before calling setState after all the async save operations.

Reply:
Got it. Thanks!

💅 
Since the API error won't catch in this `handleSubmitCallback`, no matter whether an error occurs or not in *actions.js*, `onContinue()` will be triggered and proceed to the next step, so perhaps there is no need to `setSaving( false );` here.

Reply:
> Since the API error won't catch in this `handleSubmitCallback`, no matter whether an error occurs or not in _actions.js_, `onContinue()` will be triggered and proceed to the next step

Well, clearly this is an issue. 😄 

Related to https://github.com/woocommerce/google-listings-and-ads/pull/1267#discussion_r820908402, to address the above issue, I'm thinking to remove the try/catch block in `saveShippingRate` in actions.js in https://github.com/woocommerce/google-listings-and-ads/issues/1287, so that the error would bubble up and can be caught by the component. With that, the `setSaving( false )` would be possible and meaningful. Does that sound good?

Reply:
Sounds good!

A further thought is, if that API error would be thrown to this component, then perhaps we could move `setSaving( false )` into `catch` block. The continue button could be set back to enabled only if errors occurred, otherwise, keeping it showing the loading state before bringing merchant to the next step should be fine.